File: backend/export/package_projections.py
from copy import deepcopy
from typing import Any, Callable


CONCEPT_NODE_TYPES = {
    "category",
    "concept",
    "definition",
    "entity",
    "requirement",
    "standard",
    "workspace_goal",
}
PROCESS_NODE_TYPES = {"decision", "phase", "procedure", "process", "step", "workflow"}
DEPENDENCY_NODE_TYPES = {"blocker", "dependency", "risk"}
DATA_NODE_TYPES = {"data", "data_source", "dataset", "metric", "sql_query", "table"}
ACTION_NODE_TYPES = {"action", "approval", "task", "todo", "workstream"}
DEPENDENCY_RELATIONSHIP_TYPES = {
    "blocked_by",
    "blocks",
    "dependency",
    "depends_on",
    "prerequisite",
    "requires",
}
STRUCTURAL_RELATIONSHIP_TYPES = {"", "contains", "child", "parent"}
DATA_ARTIFACT_TYPES = {"chart", "data_summary", "data_table", "sql_query", "structured_data_analysis"}
EVIDENCE_ARTIFACT_TYPES = {
    "executive_summary",
    "news_article",
    "newsletter",
    "source_set_review",
    "source_summary",
}
ACTION_ARTIFACT_TYPES = {"checklist", "implementation_plan", "project_plan", "team_roadmap"}


def concept_graph(graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    return _projection(
        graph,
        "concept_graph",
        node_filter=lambda node: _node_type(node) in CONCEPT_NODE_TYPES,
        edge_filter=lambda edge: _relationship_type(edge) in STRUCTURAL_RELATIONSHIP_TYPES,
        accepted_artifacts=accepted_artifacts,
    )


def relationship_graph(
    graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None
) -> dict[str, Any]:
    return _projection(
        graph,
        "relationship_graph",
        node_filter=lambda node: _node_type(node) != "reference",
        edge_filter=lambda edge: _relationship_type(edge) not in STRUCTURAL_RELATIONSHIP_TYPES,
        accepted_artifacts=accepted_artifacts,
    )


def process_graph(graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    return _projection(
        graph,
        "process_graph",
        node_filter=lambda node: _node_type(node) in PROCESS_NODE_TYPES,
        edge_filter=lambda edge: _relationship_type(edge)
        in {*STRUCTURAL_RELATIONSHIP_TYPES, "next", "precedes", "leads_to", "decision"},
        accepted_artifacts=accepted_artifacts,
    )


def dependency_graph(
    graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None
) -> dict[str, Any]:
    return _projection(
        graph,
        "dependency_graph",
        node_filter=lambda node: _node_type(node) in DEPENDENCY_NODE_TYPES,
        edge_filter=lambda edge: _relationship_type(edge) in DEPENDENCY_RELATIONSHIP_TYPES,
        accepted_artifacts=accepted_artifacts,
        include_edge_endpoint_nodes=True,
    )


def evidence_graph(
    graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None
) -> dict[str, Any]:
    return _projection(
        graph,
        "evidence_graph",
        node_filter=lambda node: _node_type(node) != "reference",
        edge_filter=lambda edge: True,
        accepted_artifacts=accepted_artifacts,
        artifact_filter=lambda artifact: _artifact_type(artifact) in EVIDENCE_ARTIFACT_TYPES,
    )


def data_graph(graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    return _projection(
        graph,
        "data_graph",
        node_filter=lambda node: _node_type(node) in DATA_NODE_TYPES,
        edge_filter=lambda edge: _relationship_type(edge) in {*STRUCTURAL_RELATIONSHIP_TYPES, "derived_from", "measures"},
        accepted_artifacts=accepted_artifacts,
        artifact_filter=lambda artifact: _artifact_type(artifact) in DATA_ARTIFACT_TYPES,
    )


def action_graph(graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    task_node_ids = {str(task.get("node_id", "")) for task in graph.get("tasks", []) if isinstance(task, dict)}
    return _projection(
        graph,
        "action_graph",
        node_filter=lambda node: _node_type(node) in ACTION_NODE_TYPES or str(node.get("id", "")) in task_node_ids,
        edge_filter=lambda edge: _relationship_type(edge)
        in {*STRUCTURAL_RELATIONSHIP_TYPES, *DEPENDENCY_RELATIONSHIP_TYPES, "assigned_to"},
        accepted_artifacts=accepted_artifacts,
        artifact_filter=lambda artifact: _artifact_type(artifact) in ACTION_ARTIFACT_TYPES,
    )


def connected_picture_package_projections(
    graph: dict[str, Any], accepted_artifacts: list[dict[str, Any]] | None = None
) -> dict[str, Any]:
    return {
        "concept_graph": concept_graph(graph, accepted_artifacts),
        "relationship_graph": relationship_graph(graph, accepted_artifacts),
        "process_graph": process_graph(graph, accepted_artifacts),
        "dependency_graph": dependency_graph(graph, accepted_artifacts),
        "evidence_graph": evidence_graph(graph, accepted_artifacts),
        "data_graph": data_graph(graph, accepted_artifacts),
        "action_graph": action_graph(graph, accepted_artifacts),
    }


def _projection(
    graph: dict[str, Any],
    projection_type: str,
    *,
    node_filter: Callable[[dict[str, Any]], bool],
    edge_filter: Callable[[dict[str, Any]], bool],
    accepted_artifacts: list[dict[str, Any]] | None,
    artifact_filter: Callable[[dict[str, Any]], bool] | None = None,
    include_edge_endpoint_nodes: bool = False,
) -> dict[str, Any]:
    graph_nodes = [node for node in graph.get("nodes", []) if isinstance(node, dict)]
    graph_edges = [edge for edge in graph.get("edges", []) if isinstance(edge, dict)]
    matched_edges = [edge for edge in graph_edges if edge_filter(edge)]
    edge_node_ids = {
        node_id
        for edge in matched_edges
        for node_id in (str(edge.get("source_node_id", "")), str(edge.get("target_node_id", "")))
        if node_id
    }
    selected_nodes = [
        _project_node(node, projection_type)
        for node in graph_nodes
        if node_filter(node) or (include_edge_endpoint_nodes and str(node.get("id", "")) in edge_node_ids)
    ]
    selected_node_ids = {node["id"] for node in selected_nodes}
    selected_edges = [
        _project_edge(edge, projection_type)
        for edge in matched_edges
        if str(edge.get("source_node_id", "")) in selected_node_ids
        and str(edge.get("target_node_id", "")) in selected_node_ids
    ]
    selected_artifacts = [
        _project_artifact(artifact, projection_type)
        for artifact in accepted_artifacts or []
        if isinstance(artifact, dict) and (artifact_filter is None or artifact_filter(artifact))
    ]
    evidence_gaps = [
        _evidence_gap("node", node["id"], node["title"])
        for node in selected_nodes
        if node.get("needs_review") and not node.get("source_backed")
    ]
    evidence_gaps.extend(
        _evidence_gap("artifact", artifact["id"], artifact["title"])
        for artifact in selected_artifacts
        if artifact.get("needs_review") and not artifact.get("source_backed")
    )

    return {
        "id": projection_type,
        "projection_type": projection_type,
        "workspace_id": str(graph.get("workspace", {}).get("id", "")),
        "workspace_title": str(graph.get("workspace", {}).get("title", "")),
        "nodes": selected_nodes,
        "edges": selected_edges,
        "artifacts": selected_artifacts,
        "evidence_gaps": evidence_gaps,
        "metadata": {
            "source": "connected_picture_package_projection",
            "node_count": len(selected_nodes),
            "edge_count": len(selected_edges),
            "artifact_count": len(selected_artifacts),
            "needs_review_count": len(evidence_gaps),
        },
    }


def _project_node(node: dict[str, Any], projection_type: str) -> dict[str, Any]:
    source_refs = _source_refs(node)
    source_backed = _source_backed(source_refs)
    needs_review = _needs_review(node) or not source_backed
    status = "needs_review" if needs_review else str(node.get("status") or "accepted")
    return {
        "id": str(node.get("id", "")),
        "title": str(node.get("title", "")),
        "summary": str(node.get("summary", "")),
        "node_type": _node_type(node),
        "status": status,
        "confidence": deepcopy(node.get("confidence", "")),
        "source_refs": source_refs,
        "source_backed": source_backed,
        "needs_review": needs_review,
        "metadata": {
            "source": "accepted_workspace_node",
            "projection_type": projection_type,
            "parent_id": node.get("parent_id"),
            "review_reason": "" if source_backed else "Missing source evidence.",
        },
    }


def _project_edge(edge: dict[str, Any], projection_type: str) -> dict[str, Any]:
    source_refs = _source_refs(edge)
    source_backed = _source_backed(source_refs)
    needs_review = _needs_review(edge) or not source_backed
    return {
        "id": str(edge.get("id", "")),
        "source_node_id": str(edge.get("source_node_id", "")),
        "target_node_id": str(edge.get("target_node_id", "")),
        "relationship_type": _relationship_type(edge) or "contains",
        "status": "needs_review" if needs_review else str(edge.get("review_state") or "accepted"),
        "confidence": deepcopy(edge.get("confidence", edge.get("metadata", {}).get("confidence", ""))),
        "source_refs": source_refs,
        "source_backed": source_backed,
        "needs_review": needs_review,
        "metadata": {
            **deepcopy(edge.get("metadata", {}) if isinstance(edge.get("metadata"), dict) else {}),
            "source": "accepted_workspace_edge",
            "projection_type": projection_type,
            "review_reason": "" if source_backed else "Missing source evidence.",
        },
    }


def _project_artifact(artifact: dict[str, Any], projection_type: str) -> dict[str, Any]:
    source_refs = _artifact_source_refs(artifact)
    source_backed = _source_backed(source_refs)
    needs_review = _needs_review(artifact) or not source_backed
    return {
        "id": str(artifact.get("id") or f"{_artifact_type(artifact)}-artifact"),
        "artifact_type": _artifact_type(artifact),
        "title": _artifact_title(artifact),
        "status": "needs_review" if needs_review else str(artifact.get("status") or "accepted"),
        "source_refs": source_refs,
        "source_backed": source_backed,
        "needs_review": needs_review,
        "data": deepcopy(artifact.get("data", {})),
        "metadata": {
            **deepcopy(artifact.get("metadata", {}) if isinstance(artifact.get("metadata"), dict) else {}),
            "source": "accepted_artifact",
            "projection_type": projection_type,
            "review_reason": "" if source_backed else "Missing source evidence.",
        },
    }


def _artifact_source_refs(artifact: dict[str, Any]) -> list[dict[str, Any]]:
    for source in (artifact, artifact.get("data", {}), artifact.get("metadata", {})):
        refs = source.get("source_refs") if isinstance(source, dict) else None
        if isinstance(refs, list):
            return deepcopy(refs)
    return []


def _source_refs(item: dict[str, Any]) -> list[dict[str, Any]]:
    refs = item.get("source_refs")
    return deepcopy(refs) if isinstance(refs, list) else []


def _source_backed(source_refs: list[dict[str, Any]]) -> bool:
    return any(isinstance(ref, dict) and ref.get("document_id") for ref in source_refs)


def _needs_review(item: dict[str, Any]) -> bool:
    status = str(item.get("status", "")).lower()
    review_state = str(item.get("review_state", "")).lower()
    return status == "needs_review" or review_state == "needs_review" or item.get("needs_review") is True


def _node_type(node: dict[str, Any]) -> str:
    return str(node.get("node_type") or "").lower()


def _relationship_type(edge: dict[str, Any]) -> str:
    return str(edge.get("relationship_type") or "").lower()


def _artifact_type(artifact: dict[str, Any]) -> str:
    return str(artifact.get("artifact_type") or "").lower()


def _artifact_title(artifact: dict[str, Any]) -> str:
    data = artifact.get("data", {}) if isinstance(artifact.get("data"), dict) else {}
    return str(
        artifact.get("title")
        or data.get("title")
        or data.get("headline")
        or data.get("name")
        or artifact.get("artifact_type")
        or "Accepted artifact"
    )


def _evidence_gap(item_type: str, item_id: str, title: str) -> dict[str, Any]:
    return {
        "item_type": item_type,
        "item_id": item_id,
        "title": title,
        "status": "needs_review",
        "reason": "Missing source evidence.",
    }

File: frontend/src/views/projections/packageProjections.js
import {
    getPackageGraphProjection,
    getPackageConceptMapProjection,
    getPackageProcessProjection,
    getPackageRelationshipProjection,
    hasConnectedPackageProjectionData
} from '../../connected-package/connectedPackageProjections.js';

const asArray = (value) => (Array.isArray(value) ? value.filter(Boolean) : []);

const normalizeSignal = (value = '') =>
    String(value || '')
        .trim()
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, '_')
        .replace(/^_+|_+$/g, '');

const ACCEPTED_SIGNALS = new Set(['accepted', 'approved', 'reviewed', 'published', 'source_backed']);
const ACTION_TYPES = new Set(['task', 'action', 'decision', 'milestone', 'risk', 'requirement']);
const PROCESS_TYPES = new Set(['workflow', 'procedure', 'process', 'handoff', 'phase', 'checkpoint']);
const DEPENDENCY_RELATIONSHIPS = new Set([
    'depends_on',
    'dependency',
    'requires',
    'blocked_by',
    'blocks',
    'prerequisite'
]);
const PROCESS_RELATIONSHIPS = new Set([
    ...DEPENDENCY_RELATIONSHIPS,
    'sequence',
    'next',
    'then',
    'handoff',
    'leads_to',
    'leads-to',
    'contains'
]);

const isAcceptedArtifact = (artifact = {}) =>
    Boolean(artifact.accepted) ||
    ACCEPTED_SIGNALS.has(normalizeSignal(artifact.review_state)) ||
    ACCEPTED_SIGNALS.has(normalizeSignal(artifact.status));

const acceptedArtifactsForNode = (node = {}) =>
    asArray(node.generated_artifacts).filter((artifact) => isAcceptedArtifact(artifact));

const hasAcceptedPackageContent = (node = {}) =>
    ACCEPTED_SIGNALS.has(normalizeSignal(node.review_state)) ||
    ACCEPTED_SIGNALS.has(normalizeSignal(node.status)) ||
    acceptedArtifactsForNode(node).length > 0 ||
    asArray(node.local_preview_acceptances).length > 0;

const sourceRefsFor = (node = {}, artifacts = []) => {
    const refs = [
        ...asArray(node.source_refs),
        ...artifacts.flatMap((artifact) => asArray(artifact.source_refs || artifact.data?.source_refs))
    ];
    const seen = new Set();
    return refs.filter((ref) => {
        const key = JSON.stringify(ref);
        if (seen.has(key)) {
            return false;
        }
        seen.add(key);
        return true;
    });
};

const evidenceStateFor = (sourceRefs = [], reviewState = '') =>
    sourceRefs.length > 0 ? normalizeSignal(reviewState) || 'source_backed' : 'needs_review';

const packageNode = (node = {}) => {
    const artifacts = acceptedArtifactsForNode(node);
    const source_refs = sourceRefsFor(node, artifacts);
    return {
        id: node.id,
        title: node.title,
        summary: node.summary,
        node_type: node.node_type,
        status: node.status,
        review_state: node.review_state || node.status || '',
        evidence_state: evidenceStateFor(source_refs, node.review_state || node.status),
        needs_review: source_refs.length === 0,
        source_refs,
        artifact_ids: artifacts.map((artifact) => artifact.id).filter(Boolean),
        artifacts
    };
};

const relationshipType = (edge = {}) =>
    edge.relationship_type || edge.data?.relationship_type || edge.data?.type || edge.type || '';

const packageEdge = (edge = {}, nodeLookup = new Map()) => ({
    id: edge.id || `${edge.source || ''}:${edge.target || ''}:${relationshipType(edge)}`,
    source: edge.source,
    target: edge.target,
    relationship_type: relationshipType(edge),
    label: edge.label || edge.data?.label || relationshipType(edge),
    source_refs: asArray(edge.source_refs || edge.data?.source_refs),
    source_title: nodeLookup.get(edge.source)?.title || '',
    target_title: nodeLookup.get(edge.target)?.title || ''
});

const acceptedPackageNodes = (projection = {}) =>
    asArray(projection.nodes).filter(hasAcceptedPackageContent).map(packageNode);

const acceptedNodeLookup = (nodes = []) => new Map(nodes.map((node) => [node.id, node]));

const acceptedPackageEdges = (projection = {}, nodes = acceptedPackageNodes(projection)) => {
    const nodeLookup = acceptedNodeLookup(nodes);
    return asArray(projection.edges)
        .filter((edge) => nodeLookup.has(edge.source) && nodeLookup.has(edge.target))
        .map((edge) => packageEdge(edge, nodeLookup));
};

const graphProjection = (projection, type, nodes, edges, extra = {}) => ({
    projection_type: type,
    eligible: nodes.length > 0 || edges.length > 0,
    node_count: nodes.length,
    edge_count: edges.length,
    needs_review_count: nodes.filter((node) => node.needs_review).length,
    nodes,
    edges,
    ...extra
});

export const getPackageReadyProjection = (projection = {}) => {
    if (hasConnectedPackageProjectionData(projection)) {
        return getPackageGraphProjection(projection, { projectionType: 'package_ready' });
    }
    const nodes = acceptedPackageNodes(projection);
    const edges = acceptedPackageEdges(projection, nodes);
    return graphProjection(projection, 'package_ready', nodes, edges);
};

export const getConceptGraphProjection = (projection = {}) => {
    if (hasConnectedPackageProjectionData(projection)) {
        return getPackageConceptMapProjection(projection);
    }
    const nodes = acceptedPackageNodes(projection).filter(
        (node) => !ACTION_TYPES.has(normalizeSignal(node.node_type))
    );
    return graphProjection(projection, 'concept_graph', nodes, acceptedPackageEdges(projection, nodes));
};

export const getRelationshipGraphProjection = (projection = {}) => {
    if (hasConnectedPackageProjectionData(projection)) {
        return getPackageRelationshipProjection(projection);
    }
    const nodes = acceptedPackageNodes(projection);
    const edges = acceptedPackageEdges(projection, nodes).filter((edge) =>
        Boolean(edge.relationship_type)
    );
    return graphProjection(projection, 'relationship_graph', nodes, edges);
};

export const getProcessGraphProjection = (projection = {}) => {
    if (hasConnectedPackageProjectionData(projection)) {
        return getPackageProcessProjection(projection);
    }
    const nodes = acceptedPackageNodes(projection).filter((node) =>
        PROCESS_TYPES.has(normalizeSignal(node.node_type))
    );
    const edges = acceptedPackageEdges(projection, nodes).filter((edge) =>
        PROCESS_RELATIONSHIPS.has(normalizeSignal(edge.relationship_type))
    );
    return graphProjection(projection, 'process_graph', nodes, edges);
};

export const getDependencyGraphProjection = (projection = {}) => {
    const nodes = acceptedPackageNodes(projection);
    const edges = acceptedPackageEdges(projection, nodes).filter((edge) =>
        DEPENDENCY_RELATIONSHIPS.has(normalizeSignal(edge.relationship_type))
    );
    const dependencyNodeIds = new Set(edges.flatMap((edge) => [edge.source, edge.target]));
    return graphProjection(
        projection,
        'dependency_graph',
        nodes.filter((node) => dependencyNodeIds.has(node.id)),
        edges
    );
};

export const getEvidenceGraphProjection = (projection = {}) => {
    const nodes = acceptedPackageNodes(projection);
    const evidenceNodes = nodes.flatMap((node) =>
        node.source_refs.length
            ? node.source_refs.map((ref, index) => ({
                  id:
                      ref.source_id ||
                      ref.document_id ||
                      ref.file_id ||
                      `${node.id}:source:${index}`,
                  title: ref.title || ref.filename || ref.document_id || ref.source_id || 'Source',
                  source_ref: ref,
                  evidence_state: 'source_backed'
              }))
            : [
                  {
                      id: `${node.id}:missing-evidence`,
                      title: 'Missing evidence',
                      source_ref: null,
                      evidence_state: 'needs_review'
                  }
              ]
    );
    const edges = nodes.map((node) => ({
        id: `${node.id}:evidence`,
        source: node.source_refs[0]?.source_id || node.source_refs[0]?.document_id || `${node.id}:missing-evidence`,
        target: node.id,
        relationship_type: 'supports',
        evidence_state: node.evidence_state
    }));
    return graphProjection(projection, 'evidence_graph', [...evidenceNodes, ...nodes], edges);
};

export const getDataGraphProjection = (projection = {}) => {
    const nodes = acceptedPackageNodes(projection).filter((node) =>
        node.artifacts.some((artifact) =>
            ['data_table', 'chart', 'sql_query'].includes(normalizeSignal(artifact.artifact_type))
        )
    );
    return graphProjection(projection, 'data_graph', nodes, acceptedPackageEdges(projection, nodes), {
        artifacts: nodes.flatMap((node) => node.artifacts)
    });
};

export const getActionGraphProjection = (projection = {}) => {
    const nodes = acceptedPackageNodes(projection).filter((node) =>
        ACTION_TYPES.has(normalizeSignal(node.node_type))
    );
    return graphProjection(projection, 'action_graph', nodes, acceptedPackageEdges(projection, nodes));
};
